From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Nagarjuna Kristam <nkristam@nvidia.com>
Date: Mon, 14 Nov 2022 10:48:53 +0530
Subject: [PATCH 220/254] NVIDIA: SAUCE: PCI: designware-host: Update resource
 cleanup

Config space is mapped on every host init, however, its free'ed
only when corresponding device is unbinded. During hotplug, host
init is called multiple times causing mapping issues. So map
config space only if its not mapped previously.

Similarly during deinit, bridge resources are not freed, release the
same.

Bug 3868928

Change-Id: I7b0af900f1b4eac36836a8a43ef9ac399bd9312c
Signed-off-by: Nagarjuna Kristam <nkristam@nvidia.com>
Reviewed-by: Manikanta Maddireddy <mmaddireddy@nvidia.com>
Reviewed-by: Bitan Biswas <bbiswas@nvidia.com>
Reviewed-on: https://git-master.nvidia.com/r/c/linux-stable/+/2897676
Tested-by: Abhilash G <abhilashg@nvidia.com>
Reviewed-by: Abhilash G <abhilashg@nvidia.com>
Reviewed-by: Laxman Dewangan <ldewangan@nvidia.com>

diff --git a/drivers/pci/controller/dwc/pcie-designware-host.c b/drivers/pci/controller/dwc/pcie-designware-host.c
index 20778b85afde5..8a36bf44cb329 100644
--- a/drivers/pci/controller/dwc/pcie-designware-host.c
+++ b/drivers/pci/controller/dwc/pcie-designware-host.c
@@ -708,6 +708,8 @@ EXPORT_SYMBOL_GPL(dw_pcie_host_init);
 void dw_pcie_host_deinit(struct dw_pcie_rp *pp)
 {
 	struct dw_pcie *pci = to_dw_pcie_from_pp(pp);
+	struct device *dev = pci->dev;
+	struct resource_entry *win, *tmp;
 
 	dwc_pcie_debugfs_deinit(pci);
 
@@ -726,6 +728,21 @@ void dw_pcie_host_deinit(struct dw_pcie_rp *pp)
 
 	if (pp->cfg)
 		pci_ecam_free(pp->cfg);
+
+	resource_list_for_each_entry_safe(win, tmp, &pp->bridge->windows) {
+		switch (resource_type(win->res)) {
+		case IORESOURCE_IO:
+			pci_unmap_iospace(win->res);
+			devm_release_resource(dev, win->res);
+			break;
+		case IORESOURCE_MEM:
+			devm_release_resource(dev, win->res);
+			break;
+		default:
+			continue;
+		}
+	}
+	pci_free_host_bridge(pp->bridge);
 }
 EXPORT_SYMBOL_GPL(dw_pcie_host_deinit);
 
-- 
2.43.0

